### PR TITLE
[DebugInfo][RemoveDIs] Instrument jump-threading to update DPValues

### PR DESCRIPTION
**Pull request by jmorse**
This patch makes jump-threading handle non-instruction debug-info stored in DPValues in the same way that it updates dbg.values nowadays. This involves re-targetting their operands as with dbg.values getting moved from one block to another, and manually cloning them when duplicating blocks. The SSAUpdater class also grows some functions for SSA-updating DPValues in the same way as dbg.values.

All of this is largely covered by existing debug-info tests, except for the cloning of DPValues attached to elidable instructions and branches, where I've added a test to thread-debug-info.ll. Where previously we could rely on dbg.values being copied and cloned as normal instructions are, as we need to explicitly perform that operation now I've added some explicit testing for it.

### CODE REVIEW

File: llvm/test/Transforms/JumpThreading/thread-debug-info.ll [hunk 1]
```diff
@@ -91,6 +92,47 @@ exit:                                             ; preds = %bb.f4, %bb.f3, %bb.
   ret void, !dbg !29
 }
 
+; Test for the cloning of dbg.values on elided instructions -- down one path
+; being threaded, the and in the function below is optimised away, but it's
```
Comment (OCHyams) [inline llvm/test/Transforms/JumpThreading/thread-debug-info.ll:96]: Could you demark "and" so it doesn't look like a word, e.g. \`and\`?

nit: it's -> its (apostrophe is possessive unless there's already a contraction, I think is the rule. At the very least, "it's" is always "it is" and never "belonging to it").

There's another "it's" in the next sentence that wants the same treatment.

File: llvm/test/Transforms/JumpThreading/thread-debug-info.ll [hunk 2]
```diff
@@ -91,6 +92,47 @@ exit:                                             ; preds = %bb.f4, %bb.f3, %bb.
...
+  br i1 %c, label %land.end, label %land.rhs, !dbg !33
+
+land.rhs:
+  br i1 %c2, label %lor.lhs.false.i, label %land.end, !dbg !33
+
+lor.lhs.false.i:
+  br i1 %c3, label %land.end, label %land.end, !dbg !33
+
+; CHECK-LABEL: land.end.thr_comm:
+; CHECK-NEXT:  call void @llvm.dbg.value(metadata i32 0,
```
Comment (OCHyams) [inline llvm/test/Transforms/JumpThreading/thread-debug-info.ll:112]: Can you give the dbg.values different constant values to easily identify which is which in the output/CHECK lines?

**Review by OCHyams — APPROVED** [review]
LGTM with some nits in the test.

File: llvm/lib/Transforms/Scalar/JumpThreading.cpp [hunk 3]
```diff
@@ -410,6 +410,10 @@ static bool replaceFoldableUses(Instruction *Cond, Value *ToVal,
   if (Cond->getParent() == KnownAtEndOfBB)
     Changed |= replaceNonLocalUsesWith(Cond, ToVal);
   for (Instruction &I : reverse(*KnownAtEndOfBB)) {
+    // Replace any debug-info record users of Cond with ToVal.
+    for (DPValue &DPV : I.getDbgValueRange())
+      DPV.replaceVariableLocationOp(Cond, ToVal, true);
```
[the remark this answers is not in the document]
Comment (jmorse) [reply llvm/lib/Transforms/Scalar/JumpThreading.cpp:415]: It's more of a check-my-errors flag -- using replaceVariableLocationOp on debug intrinsics that don't use the "Source" value is considered an error, and there's an assertion for that. Wheras this code here is speculatively trying to "replace this value if it's used, never mind if it isn't", and the AllowEmpty parameter signals that there shouldn't be an error if the Source value can't be found.

There are a bunch of different ways of doing this, I'll admit I haven't thought about how to restructure it, but it's an existing API call on DbgValueInst, so I didn't want to get too far away from that.

**Pull request merged**